### PR TITLE
FIX: Dismiss unread posts in subcategories

### PR DESCRIPTION
**Pull request by pmusaraj**
When a parent category shows topics from subcategories, dismissing should dismiss posts in both parent and subcategories.

See meta t/335195 for the original report.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/controllers/discovery/list.js [hunk 1]
```diff
@@ -207,7 +207,7 @@ export default class DiscoveryListController extends Controller {
     this.bulkSelectHelper.dismissRead(operationType, {
       categoryId: this.model.category?.id,
       tagName: this.model.tag?.id,
-      includeSubcategories: this.model.noSubcategories,
+      includeSubcategories: !this.model.noSubcategories,
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/controllers/discovery/list.js:210]: It feels a bit counter-intuitive to use this, but I think it's the correct thing to do. When `noSubcategories=true`, topics from subcategories aren't listed, so we shouldn't dismiss them. Vice-versa, when it is false, topics from subcategories are shown, so they should be included in the list of topics to dismiss.

File: spec/system/category_topics_spec.rb [hunk 2]
```diff
@@ -32,3 +32,24 @@
     expect(category_list).to have_no_new_posts_badge
   end
 end
+
+describe "Viewing category topic list", type: :system do
+  let(:topic_list) { PageObjects::Components::TopicList.new }
+
+  context "when parent category has default_list_filter=none" do
```
Comment (pmusaraj) [inline spec/system/category_topics_spec.rb:39]: This spec is not related to the fix. But it is adjacent, and I noticed that we had no spec for the `default_list_filter` behaviour.

File: spec/system/category_topics_spec.rb [hunk 3]
```diff
@@ -32,3 +32,24 @@
     expect(category_list).to have_no_new_posts_badge
   end
 end
+
+describe "Viewing category topic list", type: :system do
```
Comment (tgxworld) [inline spec/system/category_topics_spec.rb:36]: I don't think it is common practice to have multiple top level describe blocks in a single system spec file. Can you see if you can combine it with the one above? Maybe something like:

```
describe "Viewing category page", type: :system do
  describe "when viewing top topics" do
    ...
  end

  describe "when viewing a parent category with `default_list_filter` set to `none`"
   end
end
```

File: spec/system/dismissing_new_spec.rb [hunk 4]
```diff
@@ -53,6 +53,26 @@
 
       using_session(:tab_1) { expect(topic_view).to have_tracking_status("regular") }
     end
+
+    context "with subcategories" do
```
Comment (tgxworld) [inline spec/system/dismissing_new_spec.rb:57]:
```suggestion
    context "when viewing a category's topic list" do
```

File: spec/system/dismissing_new_spec.rb [hunk 5]
```diff
@@ -53,6 +53,26 @@
...
+      fab!(:category) { Fabricate(:category_with_definition) }
+      fab!(:subcategory) { Fabricate(:category_with_definition, parent_category: category) }
+      fab!(:subcategory_topic) { Fabricate(:topic, category: subcategory, user: user) }
+      fab!(:subcategory_post1) { create_post(user: user, topic: subcategory_topic) }
+      fab!(:subcategory_post2) { create_post(topic: subcategory_topic) }
+
+      it "should dismiss unread posts in subcategories when they are included in the parent category topic list" do
```
[the remark this answers is not in the document]
Comment (tgxworld) [reply spec/system/dismissing_new_spec.rb:64]: I think the test will be more comprehensive if we include unread posts in the top level category as well.

**Review by tgxworld — APPROVED** [review]
The fix and tests looks good to me. Some comments about how we describe system specs but nothing major.

File: spec/system/dismissing_new_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (pmusaraj) [reply spec/system/dismissing_new_spec.rb]: Yes, good idea, done!

**Comment by pmusaraj** [comment]
Thanks @tgxworld
